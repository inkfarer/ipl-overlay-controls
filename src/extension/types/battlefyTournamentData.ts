import { Match } from './battlefyStage';

export interface Rules {
    complete: string;
    critical: string;
}

export interface CustomField {
    name: string;
    public: boolean;
    _id: string;
}

export interface Contact {
    type: string;
    details: string;
}

export interface EmailsSent {
    oneDay: boolean;
    now: boolean;
    pendingTeamNotification: boolean;
}

export interface Series {
    round: number;
    numGames: number;
    roundType: string;
    roundText: string;
}

export interface Bracket {
    tiebreakerMethod: string;
    seriesStyle: string;
    series: Series[];
    type: string;
    roundsCount: number;
    teamsCount: number;
    hasThirdPlaceMatch: boolean;
    currentRoundNumber: number;
    style: string;
}

export interface Stage {
    _id: string;
    name: string;
    startTime: Date;
    bracket: Bracket;
    hasStarted: boolean;
    hasMatchCheckin?: boolean;
    hasCheckinTimer?: boolean;
    matches?: Match[];
}

export interface Features {
    ladder: boolean;
}

export interface Owner {
    _id: string;
    timezone: string;
}

export interface Organization {
    _id: string;
    name: string;
    ownerID: string;
    features: Features;
    slug: string;
    logoUrl: string;
    followers: number;
    bannerUrl: string;
    owner: Owner;
}

export interface Game {
    _id: string;
    name: string;
    iconUrl: string;
    imageUrl: string;
    backgroundUrl: string;
    abbreviation: string;
    inGameNameLabel: string;
    createdAt: Date;
    updatedAt: Date;
    slug: string;
    playersPerTeam: number;
    tournamentCreateOrder: number;
    defaultPlayersPerTeam: number;
    cloudSearchDocumentHash: string;
    cloudSearchDocumentLastGenerated: Date;
    searchScoreMultiplier: number;
}

export interface Stream {
    _id: string;
    provider: string;
    name: string;
    link: string;
    createdAt: Date;
    updatedAt: Date;
}

export interface BattlefyTournamentData {
    _id: string;
    startTime: Date;
    rules: Rules;
    playersPerTeam: number;
    customFields: CustomField[];
    userCanReport: boolean;
    name: string;
    about: string;
    bannerUrl: string;
    contact: Contact;
    contactDetails: string;
    prizes: string;
    schedule: string;
    type: string;
    checkInRequired: boolean;
    checkInStarts: string;
    hasRegistrationCap: boolean;
    teamCap: number;
    countryFlagsOnBrackets: boolean;
    isPublished: boolean;
    hasMaxPlayers: boolean;
    serviceFeePercent: number;
    gameName: string;
    isFeatured: boolean;
    isPublic: boolean;
    isSuspended: boolean;
    isRosterLocked: boolean;
    registrationEnabled: boolean;
    createdAt: Date;
    updatedAt: Date;
    checkInStartTime: Date;
    slug: string;
    cloudSearchDocumentHash: string;
    cloudSearchDocumentLastGenerated: Date;
    clonedFromTournamentID: string;
    allowedCountries: boolean;
    maxPlayers: number;
    emailsSent: EmailsSent;
    organizationID: string;
    gameID: string;
    stageIDs: string[];
    streamIDs: string[];
    hasPassword: boolean;
    lastCompletedMatchAt: Date;
    templateID: string;
    stages: Stage[];
    organization: Organization;
    game: Game;
    streams: Stream[];
}
