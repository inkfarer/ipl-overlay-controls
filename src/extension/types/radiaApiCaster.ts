export interface RadiaApiCaster {
    discord_user_id: string;
    name?: string;
    twitter?: string;
    pronouns?: string;
}
