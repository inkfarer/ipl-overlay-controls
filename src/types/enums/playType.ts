export enum PlayType {
    PLAY_ALL = 'PLAY_ALL',
    BEST_OF = 'BEST_OF'
}
