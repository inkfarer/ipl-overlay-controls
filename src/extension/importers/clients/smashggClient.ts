import { Team } from 'types/team';
import axios from 'axios';
import { HighlightedMatches, TournamentData } from 'schemas';
import { TournamentDataSource } from 'types/enums/tournamentDataSource';
import {
    SmashggEntrantsResponse,
    SmashggEvent,
    SmashggEventsResponse,
    SmashggTournamentStreamQueueResponse,
    SmashggTournamentStreamQueueSlot
} from 'types/smashgg';
import isEmpty from 'lodash/isEmpty';
import { PlayTypeHelper } from '../../../helpers/enums/playTypeHelper';
import { normalizePronouns } from '../../helpers/PronounNormalizer';

export async function getSmashGGEvents(slug: string, token: string): Promise<SmashggEvent[]> {
    const query = `query Events($slug: String!) {
      tournament(slug: $slug) {
        id
        events {
          id
          name
          videogame {
            displayName
          }
        }
      }
    }`;

    const response = await axios.post<SmashggEventsResponse>(
        'https://api.smash.gg/gql/alpha',
        JSON.stringify({
            query,
            variables: {
                slug
            }
        }),
        {
            headers: {
                'Content-Type': 'application/json',
                Accept: 'application/json',
                Authorization: `Bearer ${token}`
            }
        }
    );

    if (!response.data.data.tournament) {
        throw new Error(`Could not find tournament with slug '${slug}'.`);
    }
    if (!response.data.data.tournament.events) {
        throw new Error(`Tournament '${slug}' has no events.`);
    }

    return response.data.data.tournament.events.map(event => ({
        id: event.id,
        name: event.name,
        game: event.videogame.displayName
    }));
}

export async function getSmashGGData(eventId: number, token: string): Promise<TournamentData> {
    const firstPageData = await getSmashGGPage(1, eventId, token, true);
    const slug = firstPageData.raw.data.event.tournament.slug;

    const tourneyInfo: TournamentData = {
        meta: {
            id: slug.substring(slug.indexOf('/') + 1),
            source: TournamentDataSource.SMASHGG,
            name: firstPageData.raw.data.event.tournament.name,
            shortName: firstPageData.raw.data.event.tournament.name ?? 'Unknown Tournament',
            url: `https://smash.gg/${firstPageData.raw.data.event.tournament.slug}/details`,
            sourceSpecificData: {
                smashgg: {
                    tournamentId: firstPageData.raw.data.event.tournament.id,
                    streams: firstPageData.raw.data.event.tournament.streams,
                    eventData: {
                        id: firstPageData.raw.data.event.id,
                        name: firstPageData.raw.data.event.name,
                        game: firstPageData.raw.data.event.videogame.displayName
                    }
                }
            }
        },
        teams: firstPageData.pageInfo
    };

    // If there are more pages, add them to our data set
    if (firstPageData.raw.data.event.entrants.pageInfo.totalPages > 1) {
        const pagePromises = [];
        for (let i = 2; i <= firstPageData.raw.data.event.entrants.pageInfo.totalPages; i++) {
            pagePromises.push(getSmashGGPage(i, eventId, token));
        }

        const pages = await Promise.all(pagePromises);

        for (let i = 0; i < pages.length; i++) {
            tourneyInfo.teams = tourneyInfo.teams.concat(pages[i].pageInfo);
        }
    }

    return tourneyInfo;
}

async function getSmashGGPage(
    page: number,
    eventId: number,
    token: string,
    getRaw = false
): Promise<{ pageInfo: Team[], raw?: SmashggEntrantsResponse }> {
    const query = `query EventEntrants($eventId: ID!, $page: Int!, $perPage: Int!) {
      event(id: $eventId) {
        id
        videogame {
          displayName
        }
        tournament {
          id
          name
          slug
          streams{
            id
            streamName
          }
        }
        name
        entrants(query: {
          page: $page
          perPage: $perPage
        }) {
          pageInfo {
            total
            totalPages
          }
          nodes {
            id
            name
            team {
              images {
                type
                url
              }
            }
            participants {
              id
              prefix
              gamerTag
              user{
                genderPronoun
              }
            }
          }
        }
      }
    }`;
    const perPage = '50';

    const response = await axios.post<SmashggEntrantsResponse>(
        'https://api.smash.gg/gql/alpha',
        JSON.stringify({
            query,
            variables: {
                eventId,
                page,
                perPage
            }
        }),
        {
            headers: {
                'Content-Type': 'application/json',
                Accept: 'application/json',
                Authorization: `Bearer ${token}`
            }
        }
    );

    const { data } = response;

    if (!data.data.event) {
        throw new Error(`Could not find event with id '${eventId}'`);
    }

    const pageInfo: Team[] = [];
    data.data.event.entrants.nodes.forEach(entrant => {
        pageInfo.push({
            id: entrant.id.toString(10),
            name: entrant.name,
            showLogo: true,
            logoUrl: entrant.team?.images.find(image => image.type === 'profile')?.url,
            players: entrant.participants.map(participant => ({
                name: isEmpty(participant.prefix)
                    ? participant.gamerTag : `${participant.prefix} ${participant.gamerTag}`,
                pronouns: normalizePronouns(participant.user?.genderPronoun)
            }))
        });
    });

    if (getRaw) {
        return { pageInfo, raw: data };
    } else {
        return { pageInfo };
    }
}

/**
 * Creates a Team object from SmashGG slot data
 * @param teamData Slot data for a team
 */
function streamQueueTeamCreator(teamData: SmashggTournamentStreamQueueSlot): Team {
    return {
        id: teamData.entrant.id.toString(),
        name: teamData.entrant.name,
        showLogo: true,
        seed: teamData.seed.groupSeedNum,
        players: teamData.entrant.participants.map(participant => ({
            name: isEmpty(participant.prefix)
                ? participant.gamerTag : `${participant.prefix} ${participant.gamerTag}`,
            pronouns: participant.user?.genderPronoun?.toLowerCase()
        })),
        logoUrl: teamData.entrant.team?.images?.find(image => image.type === 'profile')?.url
    };
}

/**
 * Get Queued sets
 * @param slug Tournament slug
 * @param token Smash.gg token
 * @param eventID Event ID
 * @param streamIDs List of streamID you want to get sets for.
 * @param getAllStreams Whether to fetch all streams or only the streams provided
 */
export async function getSmashGGStreamQueue(
    slug: string,
    token: string,
    eventID: number,
    streamIDs?: number[],
    getAllStreams?: boolean
): Promise<HighlightedMatches> {
    if ((!streamIDs || streamIDs.length <= 0) && !getAllStreams) {
        return [];
    }

    const query = `query EventEntrants($slug: String!) {
        tournament(slug: $slug) {
            streamQueue {
                stream{
                  id
                }
                sets {
                    id
                    identifier
                    round
                    setGamesType
                    event {
                        id
                    }
                    phaseGroup {
                        displayIdentifier
                        phase {
                            name
                        }
                    }
                    slots {
                        seed {
                            groupSeedNum
                        }
                        entrant {
                            id
                            name
                            team {
                                images {
                                    url
                                    type
                                }
                            }
                            participants {
                                prefix
                                gamerTag
                                user {
                                    genderPronoun
                                }
                            }
                        }
                    }
                }
            }
        }
    }`;

    const response = await axios.post<SmashggTournamentStreamQueueResponse>(
        'https://api.smash.gg/gql/alpha',
        JSON.stringify({
            query,
            variables: {
                slug,
            }
        }),
        {
            headers: {
                'Content-Type': 'application/json',
                Accept: 'application/json',
                Authorization: `Bearer ${token}`
            }
        }
    );

    const highlightedStreamQueue: HighlightedMatches = [];
    const { data } = response;

    if (!data.data?.tournament?.streamQueue) {
        return [];
    }

    data.data.tournament.streamQueue.forEach(streamQueue => {
        if (getAllStreams || streamIDs.includes(streamQueue.stream.id)) {
            streamQueue.sets.forEach(set => {
                // If there's 2 teams and neither of the objects are null and matched the event we imported
                if (set.slots.length === 2
                    && !(set.slots.some(slot => slot.entrant === null))
                    && set.event.id === eventID) {
                    highlightedStreamQueue.push({
                        meta: {
                            id: set.id.toString(),
                            stageName: set.phaseGroup.displayIdentifier,
                            round: set.round,
                            match: 0,
                            name: `Set ${set.identifier} - Round ${set.round} - Pool ${set.phaseGroup.displayIdentifier} - ${set.phaseGroup.phase.name}`,
                            playType: PlayTypeHelper.fromSmashggSetGamesType(set.setGamesType)
                        },
                        teamA: streamQueueTeamCreator(set.slots[0]),
                        teamB: streamQueueTeamCreator(set.slots[1])
                    });
                }
            });
        }
    });
    return highlightedStreamQueue;
}

