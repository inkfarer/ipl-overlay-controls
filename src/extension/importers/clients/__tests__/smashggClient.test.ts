import { mock } from 'jest-mock-extended';
import type * as PronounNormalizer from '../../../helpers/PronounNormalizer';
const mockPronounNormalizer = mock<typeof PronounNormalizer>();
jest.mock('../../../helpers/PronounNormalizer', () => mockPronounNormalizer);

import axios from 'axios';
import { getSmashGGData, getSmashGGEvents, getSmashGGStreamQueue } from '../smashggClient';
import { TournamentDataSource } from 'types/enums/tournamentDataSource';
import { SmashggEntrantsResponse } from 'types/smashgg';
import { PlayType } from '../../../../types/enums/playType';

describe('smashggClient', () => {
    jest.mock('axios');
    const mockPost = jest.fn();

    beforeEach(() => {
        jest.resetAllMocks();

        axios.post = mockPost;
    });

    describe('getSmashGGEvents', () => {
        it('fetches events from Smash.gg', async () => {
            mockPost.mockResolvedValue({
                data: {
                    data: {
                        tournament: {
                            events: [
                                {
                                    id: 123431423,
                                    name: 'Cool Event',
                                    videogame: {
                                        displayName: 'Splatoon 2'
                                    }
                                },
                                {
                                    id: 238470278,
                                    name: 'Cool Smash',
                                    videogame: {
                                        displayName: 'Ultimate'
                                    }
                                }
                            ]
                        }
                    }
                }
            });

            const result = await getSmashGGEvents('event', 'apdjwiadhjwio');

            expect(result).toEqual([
                {
                    id: 123431423,
                    name: 'Cool Event',
                    game: 'Splatoon 2'
                },
                {
                    id: 238470278,
                    name: 'Cool Smash',
                    game: 'Ultimate'
                }
            ]);
            expect(mockPost).toHaveBeenCalledWith(
                'https://api.smash.gg/gql/alpha',
                expect.any(String),
                {
                    headers: {
                        'Content-Type': 'application/json',
                        Accept: 'application/json',
                        Authorization: 'Bearer apdjwiadhjwio'
                    }
                }
            );
            expect(cleanUpGraphqlQuery(mockPost.mock.calls[0][1])).toMatchSnapshot();
        });

        it('throws error if tournament is not found', async () => {
            mockPost.mockResolvedValue({
                data: {
                    data: {
                        tournament: null
                    }
                }
            });

            await expect(() => getSmashGGEvents('event', 'apdjwiadhjwio'))
                .rejects.toThrow(new Error('Could not find tournament with slug \'event\'.'));
        });

        it('throws error if tournament has no events', async () => {
            mockPost.mockResolvedValue({
                data: {
                    data: {
                        tournament: {
                            events: null
                        }
                    }
                }
            });

            await expect(() => getSmashGGEvents('eventte', 'apdjwiadhjwio'))
                .rejects.toThrow(new Error('Tournament \'eventte\' has no events.'));
        });
    });

    describe('getSmashGGData', () => {
        it('fetches data from Smash.gg and turns it into our expected format', async () => {
            mockPronounNormalizer.normalizePronouns.mockImplementation((input) => input);
            mockPost.mockResolvedValue({
                data: {
                    data: {
                        event: {
                            id: 1234567,
                            name: 'Splatoon Two',
                            videogame: {
                                displayName: 'Splatoon 2'
                            },
                            tournament: {
                                name: 'Cool Tournament',
                                id: 654423,
                                slug: 'ct',
                                streams: [
                                    { id: 994422, streamName: 'Cool Stream' }
                                ]
                            },
                            entrants: {
                                pageInfo: {
                                    total: 10,
                                    totalPages: 1
                                },
                                nodes: [
                                    {
                                        id: 1111111,
                                        name: 'Team One',
                                        team: {
                                            images: [
                                                {
                                                    url: 'smashgg://img',
                                                    type: 'profile'
                                                }
                                            ]
                                        },
                                        participants: [
                                            {
                                                id: 12345,
                                                prefix: 'TO',
                                                gamerTag: 'Player One',
                                                user: {
                                                    genderPronoun: 'he/him'
                                                }
                                            },
                                            {
                                                id: 23456,
                                                gamerTag: 'Player Two',
                                                user: { }
                                            }
                                        ]
                                    },
                                    {
                                        id: 2222222,
                                        name: 'Team Two',
                                        participants: [
                                            {
                                                id: 45678,
                                                gamerTag: 'Player Three',
                                                user: { }
                                            },
                                            {
                                                id: 23456,
                                                prefix: 'TT',
                                                gamerTag: 'Player Four',
                                                user: {
                                                    genderPronoun: 'She/Her'
                                                }
                                            }
                                        ]
                                    }
                                ]
                            }
                        }
                    }
                } as SmashggEntrantsResponse
            });

            const result = await getSmashGGData(123123, '149083257830574');

            expect(result).toEqual({
                meta: {
                    id: 'ct',
                    name: 'Cool Tournament',
                    shortName: 'Cool Tournament',
                    source: TournamentDataSource.SMASHGG,
                    sourceSpecificData: {
                        smashgg: {
                            eventData: {
                                game: 'Splatoon 2',
                                id: 1234567,
                                name: 'Splatoon Two'
                            },
                            streams: [
                                { id: 994422, streamName: 'Cool Stream' }
                            ],
                            tournamentId: 654423
                        }
                    },
                    url: 'https://smash.gg/ct/details'
                },
                teams: [
                    {
                        id: '1111111',
                        name: 'Team One',
                        logoUrl: 'smashgg://img',
                        showLogo: true,
                        players: [
                            { name: 'TO Player One', pronouns: 'he/him' },
                            { name: 'Player Two', pronouns: undefined }
                        ]
                    },
                    {
                        id: '2222222',
                        name: 'Team Two',
                        showLogo: true,
                        logoUrl: undefined,
                        players: [
                            { name: 'Player Three', pronouns: undefined },
                            { name: 'TT Player Four', pronouns: 'She/Her' }
                        ]
                    }
                ]
            });
            expect(mockPost).toHaveBeenCalledWith(
                'https://api.smash.gg/gql/alpha',
                expect.any(String),
                {
                    headers: {
                        'Content-Type': 'application/json',
                        Accept: 'application/json',
                        Authorization: 'Bearer 149083257830574'
                    }
                }
            );
            expect(cleanUpGraphqlQuery(mockPost.mock.calls[0][1])).toMatchSnapshot();
            expect(mockPronounNormalizer.normalizePronouns).toHaveBeenCalledWith('She/Her');
            expect(mockPronounNormalizer.normalizePronouns).toHaveBeenCalledWith('he/him');
            expect(mockPronounNormalizer.normalizePronouns).toHaveBeenCalledWith(undefined);
        });

        it('fetches data multiple times if more than one page is available', async () => {
            mockPronounNormalizer.normalizePronouns.mockImplementation((input) => input);
            mockPost.mockResolvedValueOnce({
                data: {
                    data: {
                        event: {
                            id: 1234567,
                            name: 'Splatoon Two',
                            videogame: {
                                displayName: 'Splatoon 2'
                            },
                            tournament: {
                                name: 'Cool Tournament',
                                id: 654423,
                                slug: 'ct'
                            },
                            entrants: {
                                pageInfo: {
                                    total: 10,
                                    totalPages: 3
                                },
                                nodes: [
                                    {
                                        id: 1111111,
                                        name: 'Team One',
                                        team: {
                                            images: [
                                                {
                                                    url: 'smashgg://img',
                                                    type: 'profile'
                                                }
                                            ]
                                        },
                                        participants: [
                                            {
                                                id: 12345,
                                                prefix: 'TO',
                                                gamerTag: 'Player One',
                                                user: {
                                                    genderPronoun: 'he/him'
                                                }
                                            },
                                            {
                                                id: 23456,
                                                gamerTag: 'Player Two',
                                                user: { }
                                            }
                                        ]
                                    }
                                ]
                            }
                        }
                    }
                } as SmashggEntrantsResponse
            }).mockResolvedValueOnce({
                data: {
                    data: {
                        event: {
                            id: 1234567,
                            name: 'Splatoon Two',
                            videogame: {
                                displayName: 'Splatoon 2'
                            },
                            tournament: {
                                name: 'Cool Tournament',
                                id: 654423,
                                slug: 'ct'
                            },
                            entrants: {
                                pageInfo: {
                                    total: 10,
                                    totalPages: 3
                                },
                                nodes: [
                                    {
                                        id: 345345345,
                                        name: 'Cool Team',
                                        team: {
                                            images: [
                                                {
                                                    url: 'smashgg://img_cool',
                                                    type: 'profile'
                                                }
                                            ]
                                        },
                                        participants: [
                                            {
                                                id: 12345,
                                                prefix: 'CT',
                                                gamerTag: 'Player One',
                                                user: {
                                                    genderPronoun: 'he/him'
                                                }
                                            },
                                            {
                                                id: 23456,
                                                gamerTag: 'Player Two',
                                                user: { }
                                            }
                                        ]
                                    },
                                    {
                                        id: 56756757,
                                        name: 'Team Three',
                                        team: {
                                            images: []
                                        },
                                        participants: [
                                            {
                                                id: 45678,
                                                gamerTag: 'Player Three',
                                                user: { }
                                            },
                                            {
                                                id: 23456,
                                                prefix: 'TT',
                                                gamerTag: 'Player Four',
                                                user: {
                                                    genderPronoun: 'She/Her'
                                                }
                                            }
                                        ]
                                    }
                                ]
                            }
                        }
                    }
                } as SmashggEntrantsResponse
            }).mockResolvedValueOnce({
                data: {
                    data: {
                        event: {
                            id: 1234567,
                            name: 'Splatoon Two',
                            videogame: {
                                displayName: 'Splatoon 2'
                            },
                            tournament: {
                                name: 'Cool Tournament',
                                id: 654423,
                                slug: 'ct'
                            },
                            entrants: {
                                pageInfo: {
                                    total: 10,
                                    totalPages: 3
                                },
                                nodes: [
                                    {
                                        id: 89089,
                                        name: 'Team Four',
                                        team: {
                                            images: []
                                        },
                                        participants: [
                                            {
                                                id: 45678,
                                                gamerTag: 'Player Three',
                                                user: { }
                                            },
                                            {
                                                id: 23456,
                                                prefix: 'TT',
                                                gamerTag: 'Player Four',
                                                user: {
                                                    genderPronoun: 'She/Her'
                                                }
                                            }
                                        ]
                                    }
                                ]
                            }
                        }
                    }
                } as SmashggEntrantsResponse
            });

            const result = await getSmashGGData(1234, '149083257830574');

            expect(result).toEqual({
                meta: {
                    id: 'ct',
                    name: 'Cool Tournament',
                    shortName: 'Cool Tournament',
                    source: TournamentDataSource.SMASHGG,
                    sourceSpecificData: {
                        smashgg: {
                            eventData: {
                                game: 'Splatoon 2',
                                id: 1234567,
                                name: 'Splatoon Two'
                            },
                            streams: undefined,
                            tournamentId: 654423
                        }
                    },
                    url: 'https://smash.gg/ct/details'
                },
                teams: [
                    {
                        id: '1111111',
                        name: 'Team One',
                        showLogo: true,
                        logoUrl: 'smashgg://img',
                        players: [
                            { name: 'TO Player One', pronouns: 'he/him' },
                            { name: 'Player Two', pronouns: undefined }
                        ]
                    },
                    {
                        id: '345345345',
                        name: 'Cool Team',
                        logoUrl: 'smashgg://img_cool',
                        showLogo: true,
                        players: [
                            { name: 'CT Player One', pronouns: 'he/him' },
                            { name: 'Player Two', pronouns: undefined }
                        ]
                    },
                    {
                        id: '56756757',
                        name: 'Team Three',
                        logoUrl: undefined,
                        showLogo: true,
                        players: [
                            { name: 'Player Three', pronouns: undefined },
                            { name: 'TT Player Four', pronouns: 'She/Her' }
                        ]
                    },
                    {
                        id: '89089',
                        name: 'Team Four',
                        logoUrl: undefined,
                        showLogo: true,
                        players: [
                            { name: 'Player Three', pronouns: undefined },
                            { name: 'TT Player Four', pronouns: 'She/Her' }
                        ]
                    }
                ]
            });
            expect(mockPost).toHaveBeenCalledTimes(3);
            expect(mockPronounNormalizer.normalizePronouns).toHaveBeenCalledWith('She/Her');
            expect(mockPronounNormalizer.normalizePronouns).toHaveBeenCalledWith('he/him');
            expect(mockPronounNormalizer.normalizePronouns).toHaveBeenCalledWith(undefined);
        });

        it('throws error if event is not found', async () => {
            mockPost.mockResolvedValueOnce({
                data: {
                    data: {
                        event: null
                    }
                }
            });

            await expect(() => getSmashGGData(12345, '149083257830574'))
                .rejects.toEqual(new Error('Could not find event with id \'12345\''));
        });
    });

    describe('getSmashGGStreamQueue', () => {
        it('returns empty array if no streams are requested', async () => {
            await expect(getSmashGGStreamQueue('slugslug', 'tokenboken', 111, [], false)).resolves.toEqual([]);
            expect(mockPost).not.toHaveBeenCalled();
        });

        it('returns empty array if smash.gg returns no stream queues', async () => {
            mockPost.mockResolvedValue({
                data: {
                    data: {
                        tournament: {
                            streamQueue: null
                        }
                    }
                }
            });

            const result = await getSmashGGStreamQueue('slugslug', 'tokenboken', 1234, [111, 222], false);

            expect(result).toEqual([]);
        });

        it('returns requested streams in the active event', async () => {
            mockPost.mockResolvedValue({
                data: {
                    data: {
                        tournament: {
                            streamQueue: [
                                {
                                    stream: {
                                        id: 111
                                    },
                                    sets: [
                                        {
                                            id: 999,
                                            round: '999',
                                            phaseGroup: {
                                                displayIdentifier: 'ID999',
                                                phase: {
                                                    name: 'Phase X'
                                                }
                                            },
                                            setGamesType: 1,
                                            identifier: 'Z',
                                            event: {
                                                id: 1234
                                            },
                                            slots: [
                                                {
                                                    entrant: {
                                                        id: 123456,
                                                        name: 'Entrant One',
                                                        participants: [
                                                            { prefix: 'EO', gamerTag: 'Player One' },
                                                            {
                                                                gamerTag: 'Player Two',
                                                                user: { genderPronoun: 'He/Him' }
                                                            }
                                                        ]
                                                    },
                                                    seed: {
                                                        groupSeedNum: 2
                                                    }
                                                },
                                                {
                                                    entrant: {
                                                        id: 234234,
                                                        name: 'Entrant Two',
                                                        participants: [
                                                            { prefix: 'ET', gamerTag: 'Player One' },
                                                            {
                                                                gamerTag: 'Player Two',
                                                                user: { genderPronoun: 'She/Her' }
                                                            }
                                                        ],
                                                        team: {
                                                            images: [
                                                                {
                                                                    type: 'profile',
                                                                    url: 'smashgg://img'
                                                                }
                                                            ]
                                                        }
                                                    },
                                                    seed: {
                                                        groupSeedNum: 3
                                                    }
                                                }
                                            ]
                                        },
                                        {
                                            id: 919,
                                            round: '919',
                                            phaseGroup: {
                                                displayIdentifier: 'ID919',
                                                phase: {
                                                    name: 'Phase H'
                                                }
                                            },
                                            setGamesType: 2,
                                            identifier: 'G',
                                            event: {
                                                id: 12345
                                            },
                                            slots: [
                                                {
                                                    entrant: {
                                                        id: 123456,
                                                        name: 'Entrant One',
                                                        participants: [
                                                            { prefix: 'EO', gamerTag: 'Player One' },
                                                            {
                                                                gamerTag: 'Player Two',
                                                                user: { genderPronoun: 'He/Him' }
                                                            }
                                                        ]
                                                    },
                                                    seed: {
                                                        groupSeedNum: 2
                                                    }
                                                },
                                                {
                                                    entrant: {
                                                        id: 234234,
                                                        name: 'Entrant Two',
                                                        participants: [
                                                            { prefix: 'ET', gamerTag: 'Player One' },
                                                            {
                                                                gamerTag: 'Player Two',
                                                                user: { genderPronoun: 'She/Her' }
                                                            }
                                                        ],
                                                        team: {
                                                            images: [
                                                                {
                                                                    type: 'profile',
                                                                    url: 'smashgg://img'
                                                                }
                                                            ]
                                                        }
                                                    },
                                                    seed: {
                                                        groupSeedNum: 3
                                                    }
                                                }
                                            ]
                                        },
                                        {
                                            slots: [{ }]
                                        }
                                    ]
                                },
                                {
                                    stream: {
                                        id: 222
                                    },
                                    sets: [
                                        {
                                            id: 988,
                                            round: '987',
                                            identifier: 'I',
                                            setGamesType: 2,
                                            phaseGroup: {
                                                displayIdentifier: 'ID988',
                                                phase: {
                                                    name: 'Phase R'
                                                }
                                            },
                                            event: {
                                                id: 1234
                                            },
                                            slots: [
                                                {
                                                    entrant: {
                                                        id: 567567,
                                                        name: 'Entrant One',
                                                        participants: [
                                                            { prefix: 'EO', gamerTag: 'Player One' },
                                                            {
                                                                gamerTag: 'Player Two',
                                                                user: { genderPronoun: 'He/Him' }
                                                            }
                                                        ]
                                                    },
                                                    seed: {
                                                        groupSeedNum: 2
                                                    }
                                                },
                                                {
                                                    entrant: {
                                                        id: 789789,
                                                        name: 'Entrant Two',
                                                        participants: [
                                                            { prefix: 'ET', gamerTag: 'Player One' },
                                                            {
                                                                gamerTag: 'Player Two',
                                                                user: { genderPronoun: 'She/Her' }
                                                            }
                                                        ],
                                                        team: {
                                                            images: [
                                                                {
                                                                    type: 'profile',
                                                                    url: 'smashgg://img'
                                                                }
                                                            ]
                                                        }
                                                    },
                                                    seed: {
                                                        groupSeedNum: 3
                                                    }
                                                }
                                            ]
                                        },
                                        {
                                            slots: [{ }]
                                        }
                                    ]
                                },
                                {
                                    stream: {
                                        id: 333
                                    }
                                }
                            ]
                        }
                    }
                }
            });

            const result = await getSmashGGStreamQueue('slugslug', 'tokenboken', 1234, [111, 222], false);

            expect(result).toEqual([{
                meta: {
                    id: '999',
                    match: 0,
                    playType: PlayType.BEST_OF,
                    name: 'Set Z - Round 999 - Pool ID999 - Phase X',
                    round: '999',
                    stageName: 'ID999'
                },
                teamA: {
                    id: '123456',
                    logoUrl: undefined,
                    name: 'Entrant One',
                    players: [
                        {
                            name: 'EO Player One',
                            pronouns: undefined
                        },
                        {
                            name: 'Player Two',
                            pronouns: 'he/him'
                        }
                    ],
                    seed: 2,
                    showLogo: true
                },
                teamB: {
                    id: '234234',
                    logoUrl: 'smashgg://img',
                    name: 'Entrant Two',
                    players: [
                        {
                            name: 'ET Player One',
                            pronouns: undefined
                        },
                        {
                            name: 'Player Two',
                            pronouns: 'she/her'
                        }
                    ],
                    seed: 3,
                    showLogo: true
                }
            }, {
                meta: {
                    id: '988',
                    match: 0,
                    playType: PlayType.PLAY_ALL,
                    name: 'Set I - Round 987 - Pool ID988 - Phase R',
                    round: '987',
                    stageName: 'ID988'
                },
                teamA: {
                    id: '567567',
                    logoUrl: undefined,
                    name: 'Entrant One',
                    players: [
                        {
                            name: 'EO Player One',
                            pronouns: undefined
                        },
                        {
                            name: 'Player Two',
                            pronouns: 'he/him'
                        }
                    ],
                    seed: 2,
                    showLogo: true
                },
                teamB: {
                    id: '789789',
                    logoUrl: 'smashgg://img',
                    name: 'Entrant Two',
                    players: [
                        {
                            name: 'ET Player One',
                            pronouns: undefined
                        },
                        {
                            name: 'Player Two',
                            pronouns: 'she/her'
                        }
                    ],
                    seed: 3,
                    showLogo: true
                }
            }]);
            expect(mockPost).toHaveBeenCalledWith(
                'https://api.smash.gg/gql/alpha',
                expect.any(String),
                {
                    headers: {
                        'Content-Type': 'application/json',
                        Accept: 'application/json',
                        Authorization: 'Bearer tokenboken'
                    }
                }
            );
            expect(cleanUpGraphqlQuery(mockPost.mock.calls[0][1])).toMatchSnapshot();
        });
    });
});

function cleanUpGraphqlQuery(query: string): string {
    return query
        .replace(/\\n/g, '')
        .replace(/ {4,}/g, ' ');
}
