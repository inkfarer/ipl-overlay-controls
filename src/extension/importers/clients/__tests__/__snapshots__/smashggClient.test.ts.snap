// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`smashggClient getSmashGGData fetches data from Smash.gg and turns it into our expected format 1`] = `"{"query":"query EventEntrants($eventId: ID!, $page: Int!, $perPage: Int!) { event(id: $eventId) { id videogame { displayName } tournament { id name slug streams{ id streamName } } name entrants(query: { page: $page perPage: $perPage }) { pageInfo { total totalPages } nodes { id name team { images { type url } } participants { id prefix gamerTag user{ genderPronoun } } } } } }","variables":{"eventId":123123,"page":1,"perPage":"50"}}"`;

exports[`smashggClient getSmashGGEvents fetches events from Smash.gg 1`] = `"{"query":"query Events($slug: String!) { tournament(slug: $slug) { id events { id name videogame { displayName } } } }","variables":{"slug":"event"}}"`;

exports[`smashggClient getSmashGGStreamQueue returns requested streams in the active event 1`] = `"{"query":"query EventEntrants($slug: String!) { tournament(slug: $slug) { streamQueue { stream{ id } sets { id identifier round setGamesType event { id } phaseGroup { displayIdentifier phase { name } } slots { seed { groupSeedNum } entrant { id name team { images { url type } } participants { prefix gamerTag user { genderPronoun } } } } } } } }","variables":{"slug":"slugslug"}}"`;
